pkg/prometheus/promcfg.go

```diff
@@ -2142,6 +2177,39 @@ func (cg *ConfigGenerator) generateK8SSDConfig(
 	}
 }
 
+func (cg *ConfigGenerator) generateRoleSelectorConfig(k8sSDConfig yaml.MapSlice, roles []string, selector metav1.LabelSelector) yaml.MapSlice {
+	selectors := []yaml.MapSlice{}
+	for _, role := range roles {
+		yml := yaml.MapSlice{
+			{
+				Key:   "role",
+				Value: role,
+			},
+		}
+
+		labelSelector := labels.SelectorFromValidatedSet(labels.Set(selector.MatchLabels))
+
+		if len(selector.MatchExpressions) > 0 {
+			for _, exp := range selector.MatchExpressions {
+				// Needs to lower because the selection expects lowercase, and the exp.Operator is CamelCase.
+				// Ignoring error because we are validating on the resource selection.
+				requirement, _ := labels.NewRequirement(exp.Key, selection.Operator(strings.ToLower(string(exp.Operator))), exp.Values)
+				labelSelector = labelSelector.Add(*requirement)
+			}
+		}
+		yml = append(yml, yaml.MapItem{Key: "label", Value: labelSelector.String()})
+		selectors = append(selectors, yml)
```

Comment:
(nit) can we reduce the number of intermediate variables?
```suggestion
		selectors = append(selectors, 
yaml.MapSlice{
			{
				Key:   "role",
				Value: role,
			},
			{
				Key:   "label",
				Value: selector,
			},
		})
```